src/agent/coverage/src/block/windows.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::collections::BTreeMap;
+use std::process::Command;
+use std::time::{Duration, Instant};
+
+use anyhow::Result;
+use debugger::{
+    dbghelp::SymInfo,
+    debugger::{BreakpointId, BreakpointType, DebugEventHandler, Debugger},
+    target::Module,
+};
+
+use crate::{AppCoverageBlocks, ModuleCoverageBlocks};
+
+pub fn record(cmd: Command) -> Result<AppCoverageBlocks> {
+    let mut handler = BlockCoverageHandler::new();
+
+    let (mut dbg, _child) = Debugger::init(cmd, &mut handler)?;
+    dbg.run(&mut handler)?;
+
+    Ok(handler.coverage)
+}
+
+pub struct BlockCoverageHandler {
+    bp_to_block: BTreeMap<BreakpointId, (usize, usize)>,
+    coverage: AppCoverageBlocks,
+    started: Instant,
+    max_duration: Duration,
+    timed_out: bool,
+}
+
+impl BlockCoverageHandler {
+    pub fn new() -> Self {
+        let coverage = AppCoverageBlocks::new();
+        let bp_to_block = BTreeMap::default();
+        let started = Instant::now();
+        let max_duration = Duration::from_secs(5);
+        let timed_out = false;
+
+        Self {
+            bp_to_block,
+            coverage,
+            max_duration,
+            started,
+            timed_out,
+        }
+    }
+
+    pub fn pc(&self, dbg: &mut Debugger) -> Result<(u64, Option<SymInfo>)> {
+        let rip = iced_x86::Register::RIP;
+        let pc = dbg.read_register_u64(rip)?;
+        let sym = dbg.get_symbol(pc).ok();
+
+        Ok((pc, sym))
+    }
+
+    fn add_module(&mut self, dbg: &mut Debugger, module: &Module) {
+        let bitset = crate::pe::process_image(module.path(), false);
```

Comment:
I think you'll want to cache the module processing somehow. Disassembly is pretty fast, but we'll be doing it thousands of times without a cache and that may be noticeable.

Reply:
I'm assuming you mean cached across multiple command invocations, yeah?

Reply:
(and agreed)

Reply:
Mostly yes, but on Windows at least, a dll could be unloaded and loaded again.

Reply:
Yeah, we could have the same thing happening on Linux. Opened #285 to capture this.